spec/services/bulk_invoice_service_spec.rb

```diff
@@ -50,8 +50,11 @@
 
   describe "#orders_from" do
     it "orders with completed desc" do
-      expect(service.send(:orders_from, [1, 2]).to_sql)
-        .to include('ORDER BY completed_at DESC')
+      order_old = create(:order_with_distributor, :completed, completed_at: 2.minutes.ago)
+      order_older = create(:order_with_distributor, :completed, completed_at: 3.minutes.ago)
+
+      expect(service.send(:orders_from, [order_older.id, order_old.id]).pluck(:id))
+        .to eq([order_old.id, order_older.id])
```

Comment:
This test passes even if I remove `.order("completed_at DESC")`. The reason is that you create the orders in the same order as you expect them to be retrieved later. The database is very likely to provide rows in that order by default. That's why I suggested to create three orders and create them in an irregular sequence like:

1. old
2. oldest
3. older

To show you what I mean, I changed the test so that it covers the real sequence of rendering calls. And the test failed. I then found that my suggestion to use `find_each` overrides the `ORDER BY` clause. So I led you to code that actually doesn't solve the problem here. That's why proper testing is so important. I will add my commits to your branch for your review. If you are happy with my solution, you can squash all the commits into one.

We still have the potential performance problem if we try to render lots of invoices at once but let's leave it and see if it actually becomes a problem. I suggested to solve it because the solution seemed so simple. But we don't need to add complexity right now.

Reply:
Ah, I see you already found it. I give it as approved then @mkllnk